samza-core/src/main/java/org/apache/samza/table/remote/RemoteReadWriteTable.java

```diff
@@ -93,10 +93,30 @@ public void put(K key, V value) {
     }
 
     writeMetrics.numPuts.inc();
-    return execute(writeRateLimiter, key, value, writeFn::putAsync, writeMetrics.putNs)
-        .exceptionally(e -> {
-            throw new SamzaException("Failed to put a record with key=" + key, (Throwable) e);
-          });
+
+    final long startNs = System.nanoTime();
+    CompletableFuture<Void> ioFuture;
+    if (writeRateLimiter.isRateLimited()) {
+      ioFuture = CompletableFuture
+          .runAsync(() -> writeRateLimiter.throttle(key, value), tableExecutor)
+          .thenCompose((r) -> writeFn.putAsync(key, value));
+    } else {
+      ioFuture = writeFn.putAsync(key, value);
+    }
+    if (callbackExecutor != null) {
+      ioFuture.thenApplyAsync(r -> {
+          writeMetrics.putNs.update(System.nanoTime() - startNs);
+          return r;
+        }, callbackExecutor);
+    } else {
+      ioFuture.thenApply(r -> {
+          writeMetrics.putNs.update(System.nanoTime() - startNs);
+          return r;
+        });
+    }
+    return ioFuture.exceptionally(e -> {
```

Comment:
Quick question: how do we propagate this exception to the main runloop if the callback in ioFuture.exceptionally() was invoked by an IO thread?